src/lib/tecs/TECS.cpp

```diff
@@ -804,3 +827,75 @@ void TECS::update(float pitch, float altitude, float hgt_setpoint, float EAS_set
 	}
 }
 
+void TECS::_update_altitude_setpoint(TECSControl::Setpoint &control_setpoint, const float altitude) const
+{
+	if (_stall_detection_percentage > FLT_EPSILON) {
+		control_setpoint.altitude_reference.alt = (1.0f - _stall_detection_percentage) * control_setpoint.altitude_reference.alt
+				+ _stall_detection_percentage * altitude;
+		control_setpoint.altitude_reference.alt_rate = (1.0f - _stall_detection_percentage) *
+				control_setpoint.altitude_reference.alt_rate;
+		control_setpoint.altitude_rate_setpoint = (1.0f - _stall_detection_percentage) * control_setpoint.altitude_rate_setpoint
+				- _stall_detection_percentage * _control_param.max_sink_rate;
+	}
+}
+
+void TECS::_detect_airspeedless_stall(const float dt, const float altitude_sp, const float altitude, const float pitch,
+				      const float altitude_rate)
+{
+	if (_param_fw_stall_prv.get()) {
+		// Checks to detect an aircraft stalling: a positive altitude change is given,
+		// while the pitch command is maxed and the AV is sinking.
+		// TODO could also add airspeed rate negative, would need to filter the speed_deriv_forward.
+		if (((altitude_sp - altitude) > 0.0f) &&
+		    (pitch > 0.0f) &&
+		    (altitude_rate < 0.0f) &&
+		    (fabsf(_control_param.pitch_max - _control.getPitchSetpoint()) < FLT_EPSILON)) {
+			if (_stall_state == STALL_MODE::NO_STALL) {
+				PX4_WARN("Entered stall condition");
+				_stall_start_timestamp = hrt_absolute_time();
+				_stall_state = STALL_MODE::MAYBE_STALL;
+			}
+
+		} else if (_stall_state == STALL_MODE::MAYBE_STALL) {
+			PX4_WARN("Aborting stall");
+			_stall_state = STALL_MODE::NO_STALL;
+			_stall_start_timestamp = 0ULL;
+		}
+
+		// Only be sure to enter stall prevention, is the stal state persists for a certain amount of time.
+		if (_stall_state == STALL_MODE::MAYBE_STALL && hrt_elapsed_time(&_stall_start_timestamp) > 100_ms) {
+			PX4_WARN("Start stall reaction.");
+			_stall_state = STALL_MODE::STALL;
+		}
+
+		// End stall i we pitched the nose downwards and flying nearly at max sinkrate.
+		if ((_stall_state == STALL_MODE::STALL) && (pitch < 0.0f) && (altitude_rate < (0.5f - _control_param.max_sink_rate))) {
```

Comment:
@KonradRudin I wonder if we really need the sink rate condition. I imagine that having the nose pitched done is likely enough to break the stall. The reason I'm saying this is because there might be the risk to not get out of stall if the user has set an unfeasible max sink rate. 
Anyway, probably we will see what's needed during testing, I suggest we leave it as is for now.

Reply:
Good point. The reasoning i had is that when we end stall as soon as the nose is pitch down, we will pitch up again and might not have gained enough speed to get out of stall. Having to wait for the max sink rate might too much though.